### PR TITLE
update: persist invest/withdraw form state across tab changes

### PR DESCRIPTION
**Pull request by 1saf**
Fixes # .

Changes proposed in this pull request:
- Persist persist/withdraw form state across tab changes

**Testing Criteria**
- When adjusting the inputs on the invest and withdraw form, switch between tabs. Upon returning to each respective tab the inputs that were previously modified should persist.

**Questions**
- I've adjusted the behaviour to also persist the input changes when switching between sub-types in each form. (proportional vs custom), is this something that we want or not?

### CODE REVIEW

**Review by garethfuller — COMMENTED** [review]
When I try to access a pool page the app crashes. Not sure if it's related to these changes?

**Comment by 1saf** [comment]
@garethfuller Crashing should not happen now

File: src/components/forms/pool_actions/InvestForm.vue [hunk 1]
```diff
@@ -295,23 +297,29 @@ export default defineComponent({
...
-  emits: ['success'],
+  emits: ['success', 'update:modelValue'],
 
   props: {
     pool: { type: Object as PropType<FullPool>, required: true },
-    missingPrices: { type: Boolean, default: false }
+    missingPrices: { type: Boolean, default: false },
+    initialState: {
+      type: Object as PropType<
+        Pick<DataProps, 'investType' | 'range' | 'amounts'>
```
Comment (garethfuller) [inline src/components/forms/pool_actions/InvestForm.vue:307]: Is the `DataProps` definition missing from this component?

File: src/components/forms/pool_actions/WithdrawForm.vue [hunk 2]
```diff
@@ -214,29 +215,49 @@ export enum FormTypes {
...
+    initialState: {
+      type: Object as PropType<
+        Pick<DataProps, 'bptIn' | 'withdrawType' | 'range' | 'amounts'>
+      >,
+      default: () => ({})
+    }
   },
 
-  setup(props: { pool: FullPool }, { emit }) {
+  setup(props: { pool: FullPool; initialState }, { emit }) {
```
Comment (garethfuller) [inline src/components/forms/pool_actions/WithdrawForm.vue:249]: Same comment here as for the InvestForm

**Comment by garethfuller** [comment]
Something messed up seems to be happening with the balances on the withdraw tab. Not sure if it's caused by this PR or was already a bug.

Screenshot 2021-06-02 at 11 09 27

File: src/components/forms/pool_actions/InvestForm.vue [hunk 3]
```diff
@@ -295,23 +297,29 @@ export default defineComponent({
...
+      >,
+      default: () => ({})
+    }
   },
 
-  setup(props: { pool: FullPool }, { emit }) {
+  setup(props: { pool: FullPool; initialState }, { emit }) {
```
[the remark this answers is not in the document]
Comment (1saf) [reply src/components/forms/pool_actions/InvestForm.vue:313]: yep that is a much cleaner impl

**Comment by 1saf** [comment]
> Something messed up seems to be happening with the balances on the withdraw tab. Not sure if it's caused by this PR or was already a bug.
> 
> Screenshot 2021-06-02 at 11 09 27

hmm ill look at it

**Comment by garethfuller** [comment]
Another small bug: https://www.loom.com/share/6a8f33cb48c7411ab0a95d8bee52a220

**Comment by 1saf** [comment]
> Another small bug:
> 
> https://www.loom.com/share/6a8f33cb48c7411ab0a95d8bee52a220

Ah this is what I meant by persisting the behaviour between changing between proportional and custom inputs. So in this case we want to persist in tab changes but reset if switching between the type of invest/withdrawal?

**Comment by garethfuller** [comment]
> > Another small bug:
> > https://www.loom.com/share/6a8f33cb48c7411ab0a95d8bee52a220
> 
> Ah this is what I meant by persisting the behaviour between changing between proportional and custom inputs. So in this case we want to persist in tab changes but reset if switching between the type of invest/withdrawal?

Well it's not so much that it's the wrong behaviour, but when I switch back to proportional, notice how the percentage says 100% but the slider is still at like 50%. Happy to have persistent state between types.

One other thing to check before we merge this, test if when a large investment/withdrawal in the pool happens, the proportional amounts change. To do this you need to invest in the pool via another window.

**Comment by garethfuller** [comment]
small bug demo: https://www.loom.com/share/99b42f74debd44d3a9e3b43d5e2d485a

I don't think when you go to custom amounts, enter some values and switch back to proportional that it should maintain the custom numbers in the proportional case in anyway. There will be too many edge cases if we do that. Best just to reset the proportional case on type switch or maintain the proportional state prior to switching.

In general lets try not to overcomplicate this logic. When we get into a redesign of this flow, it shouldn't be needed.

**Pull request closed without merging**